File: WSIRegression/PMCT.java
import net.lightbody.bmp.BrowserMobProxy;
import net.lightbody.bmp.BrowserMobProxyServer;
import net.lightbody.bmp.client.ClientUtil;
import net.lightbody.bmp.core.har.Har;
import net.lightbody.bmp.core.har.HarEntry;
import net.lightbody.bmp.core.har.HarNameValuePair;
import net.lightbody.bmp.core.har.HarRequest;
import org.openqa.selenium.Proxy;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.ie.InternetExplorerDriver;
import org.openqa.selenium.remote.CapabilityType;
import org.openqa.selenium.remote.DesiredCapabilities;

import java.io.File;
import java.io.IOException;
import java.util.*;

public class PMCT {
    // allows us to filter HTTP requests down to the tags we're looking for
    private static String reqUrlFragment;
    private static Har har;
    private static BrowserMobProxy proxy;
    private static String event;
    private static HashMap<String, String> queryPairs;

    public static Expect testCase = new Expect();

    public static WebDriver start(String browser, String urlFragment) {
        // start the proxy
        proxy = new BrowserMobProxyServer();
        proxy.start(0);

        // get the Selenium proxy object
        Proxy seleniumProxy = ClientUtil.createSeleniumProxy(proxy);

        // configure it as a desired capability
        DesiredCapabilities capabilities = new DesiredCapabilities();
        capabilities.setCapability(CapabilityType.PROXY, seleniumProxy);

        WebDriver driver;

        if (Objects.equals(browser, "Chrome")) {
            driver = new ChromeDriver(capabilities);
        } else if (Objects.equals(browser, "Firefox")) {
            driver = new FirefoxDriver(capabilities);
        } else if (Objects.equals(browser, "IE")) {
            driver = new InternetExplorerDriver(capabilities);
        } else {
            driver = new FirefoxDriver(capabilities);
        }

        proxy.newHar();
        reqUrlFragment = urlFragment;

        return driver;
    }

    // used to construct english-like grammar for test cases - example: expect("v1").is("Home Page")
    public static class Expect {
        public String key;

        public void is(String val) {
            String foundValue;

            if (queryPairs.containsKey(key)) {
                foundValue = queryPairs.get(key);
                if (Objects.equals(val, foundValue)) {
                    System.out.println("Pass: '" + key + "' is '" + val + "'");
                } else {
                    System.out.println("FAIL: Expected '" + key + "' to be '" + val + "', was '" + foundValue + "' instead!");
                }
            } else {
                System.out.println("ERROR: Key '" + key + "' not found!");
                System.out.print(queryPairs);
            }
        }

        public void contains(String val) {
            String foundValue;

            if (queryPairs.containsKey(key)) {
                foundValue = queryPairs.get(key);
                if (foundValue.contains(val)) {
                    System.out.println("Pass: '" + key + "' contains '" + val + "'");
                } else {
                    System.out.println("FAIL: Expected '" + key + "' to contain '" + val + "', was '" + foundValue + "' instead!");
                }
            } else {
                System.out.println("ERROR: Key '" + key + "' not found!");
                System.out.print(queryPairs);
            }
        }
    }

    public static Expect expect(String key) {
        // get the HAR data if it hasn't been built by a previous expect call
        if (har == null) {
            boolean eventFound = false;
            har = proxy.getHar();

            List<HarEntry> entryList = har.getLog().getEntries();

            // search for the tag fire HTTP request
            for (HarEntry entry : entryList) {
                HarRequest req = entry.getRequest();
                if (req.getUrl().contains(reqUrlFragment)) {
                    queryPairs = new HashMap<>();

                    //create hashmap of querystring params to search through later
                    for (HarNameValuePair pair : req.getQueryString()) {
                        queryPairs.put(pair.getName(), pair.getValue());
                    }

                    // make sure this is the tag containing the event we're looking for
                    // if it isn't, keep looking
                    Set<String> eventSet = new HashSet<String>(Arrays.asList(queryPairs.get("events").split(",")));
                    if (eventSet.contains(event)) {
                        eventFound = true;
                        break;
                    }
                }
            }

            if (queryPairs == null) {
                System.out.println("ERROR: No requests found with url fragment '" + reqUrlFragment + "'");
            } else if (!eventFound) {
                System.out.println("WARNING: event '" + event + "' not found!");
            }
        }

        testCase.key = key;
        return testCase;
    }

    // reset stored data so we can begin logging new requests
    public static void watchFor(String tagEvent) {
        System.out.println("Watching for " + tagEvent + "...");
        har = null;
        proxy.newHar();
        queryPairs = null;
        event = tagEvent;
    }

    // call this before executing the killBrowser() command to avoid errors
    public static void stop() {
        proxy.stop();
    }

    // used for debugging
    public static void writeLog(String filename) throws IOException {
        // get the HAR data
        har = proxy.getHar();

        har.writeTo(new File(filename));

        proxy.newHar();
    }

    public static void writeLog() {
        try {
            writeLog("harlog.txt");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: WSIPython/pmct.py
from browsermobproxy import Server
from selenium import webdriver
import time

BROWSERMOB_LOCATION = r"C:\Pointmarc\QAbot\utils\browsermob-proxy-2.1.0-beta-3\bin\browsermob-proxy"
ANSI_STORE = "\x1b[93;41m"
ANSI_PASS = "\x1b[32m"
ANSI_FAIL = "\x1b[31m"
ANSI_WARNING = "\x1b[33m"
ANSI_ERROR = "\x1b[30;41m"
ANSI_CLEAR = "\x1b[0m"
beacon_url = ""


def print_pass(msg):
    print(ANSI_PASS + "Pass: " + msg + ANSI_CLEAR)


def print_fail(msg):
    print(ANSI_FAIL + "FAIL: " + msg + ANSI_CLEAR)


def print_warning(msg):
    print(ANSI_WARNING + "WARNING: " + msg + ANSI_CLEAR)


def print_error(msg):
    print(ANSI_ERROR + "ERROR: " + msg + ANSI_CLEAR)


class Expect:
    def __init__(self, key, queries, case):
        self.key = key
        self.queries = queries
        self.case = case

    def exists(self):
        if self.queries is None:
            return

        if self.key in self.queries:
            self.case.pass_test()
            print_pass("'" + self.key + "' exists: set to '" + self.queries[self.key] + "'")
        else:
            self.case.fail_test()
            print_fail("'" + self.key + "' is not set!")

    def equals(self, val):
        if self.queries is None:
            return

        if self.key in self.queries:
            if self.queries[self.key] == val:
                self.case.pass_test()
                print_pass("'" + self.key + "' is '" + val + "'")
            else:
                self.case.fail_test()
                print_fail("Expected '" + self.key + "' to be '" + val + "', was '" + self.queries[self.key] + "' instead!")
        else:
            self.case.error_test()
            print_error("Key '" + self.key + "' not found!")
            print(self.queries)

    def contains(self, val):
        if self.queries is None:
            return

        if self.key in self.queries:
            if val in self.queries[self.key]:
                self.case.pass_test()
                print_pass("'" + self.key + "' contains '" + val + "'")
            else:
                self.case.fail_test()
                print_fail("Expected '" + self.key + "' to contain '" + val + "', was '" + self.queries[self.key] + "' instead!")
        else:
            self.case.error_test()
            print_error("Key '" + self.key + "' not found!")
            print(self.queries)


class Pmct:
    def __init__(self):
        self.cases = []
        self.queries = None
        self.proxy = None
        self.beacon_url = ""
        self.watch_event = ""
        self.current_case = None

    def get_driver(self, browser, start_beacon_url):
        server = Server(BROWSERMOB_LOCATION)
        server.start()
        self.proxy = server.create_proxy()
        driver = webdriver.Firefox(proxy=self.proxy.selenium_proxy())
        self.proxy.new_har()
        self.beacon_url = start_beacon_url
        return driver

    def new_case(self, case_name):
        print("\n" + case_name)
        self.current_case = Case(case_name)
        self.cases.append(self.current_case)

    def watch_for(self, tag_event):
        print("Watching for '" + tag_event + "'...")
        self.proxy.new_har()
        self.queries = None
        self.watch_event = tag_event

    def expect(self, key):
        if self.queries is None:
            entry_list = self.proxy.har['log']['entries']
            matched_queries = [entry['request']['queryString'] for entry in entry_list if self.beacon_url in entry['request']['url']]

            if len(matched_queries) == 0:
                self.current_case.error_test()
                print_error("No requests found with url fragment '" + self.beacon_url + "'")
                return Expect(key, None)

            query_dicts = {}
            i = 0
            for query in matched_queries:
                query_dicts[i] = {}
                for the_dict in query:
                    query_dicts[i][the_dict['name']] = the_dict['value']
                i += 1

            for i, query in query_dicts.iteritems():
                if 'events' not in query:
                    continue
                events = query['events'].split(',')
                if self.watch_event in events:
                    # if self.queries is not None:
                    #     print_warning("multiple beacons fired with event '" + self.watch_event + "'!")
                    self.queries = query

            if self.queries is None:
                self.current_case.error_test()
                print_error("event '" + self.watch_event + "' not found!")

        return Expect(key, self.queries, self.current_case)

    def report(self):
        print("\nREPORT")
        for case in self.cases:
            msg = "{}: {} / {}".format(case.name, case.passed_tests, case.total_tests)
            if case.errored:
                print_error(msg)
            elif case.failed:
                print_fail(msg)
            else:
                print_pass(msg)


class Case:
    def __init__(self, name):
        self.name = name
        self._passed_tests = 0
        self._failed_tests = 0
        self._errors = 0

    def pass_test(self):
        self._passed_tests += 1

    def fail_test(self):
        self._failed_tests += 1

    def error_test(self):
        self._errors += 1

    @property
    def passed_tests(self):
        return self._passed_tests

    @property
    def total_tests(self):
        return self._passed_tests + self._failed_tests

    @property
    def failed(self):
        return self._failed_tests > 0

    @property
    def errored(self):
        return self._errors > 0
